samples/AspNetCoreMcpPerSessionTools/Program.cs

```diff
@@ -0,0 +1,111 @@
+using OpenTelemetry;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Trace;
+using AspNetCoreMcpPerSessionTools.Tools;
+using ModelContextProtocol.Server;
+
+var builder = WebApplication.CreateBuilder(args);
+
+// Register all MCP server tools - they will be filtered per session based on route
+builder.Services.AddMcpServer()
+    .WithHttpTransport(options =>
+    {
+        // Configure per-session options to filter tools based on route category
+        options.ConfigureSessionOptions = async (httpContext, mcpOptions, cancellationToken) =>
+        {
+            // Determine tool category from route parameters
+            var toolCategory = GetToolCategoryFromRoute(httpContext);
+
+            // Get the tool collection that we can modify per session
+            var toolCollection = mcpOptions.Capabilities?.Tools?.ToolCollection;
+            if (toolCollection != null)
+            {
+                // Clear all tools first
+                toolCollection.Clear();
+
+                // Add tools based on the requested category
+                switch (toolCategory?.ToLower())
+                {
+                    case "clock":
+                        // Clock category gets time/date tools
+                        AddToolsForType<ClockTool>(toolCollection);
+                        break;
+                        
+                    case "calculator":
+                        // Calculator category gets mathematical tools
+                        AddToolsForType<CalculatorTool>(toolCollection);
+                        break;
+                        
+                    case "userinfo":
+                        // UserInfo category gets session and system information tools
+                        AddToolsForType<UserInfoTool>(toolCollection);
+                        break;
+                        
+                    case "all":
+                    default:
+                        // Default or "all" category gets all tools
+                        AddToolsForType<ClockTool>(toolCollection);
+                        AddToolsForType<CalculatorTool>(toolCollection);
+                        AddToolsForType<UserInfoTool>(toolCollection);
+                        break;
+                }
+            }
+        };
+    })
+    .WithTools<ClockTool>()
+    .WithTools<CalculatorTool>()
+    .WithTools<UserInfoTool>();
```

Comment:
```suggestion
    });
```

Reply:
I think we can remove the `WithTools` calls now that we have the following logic in `ConfigureSessionOptions` which overrides anything that was previously in the ToolCollection anyway.

```csharp
                mcpOptions.Capabilities = new();
                mcpOptions.Capabilities.Tools = new();
                var toolCollection = mcpOptions.Capabilities.Tools.ToolCollection = new();
```